### PR TITLE
Let kubectl run follow rules for `--`

### LINKED ISSUE #13004: kubectl run doesn't follow rules for `--`
kubectl run --restart=Never --image=gcr.io/google_containers/busybox --command -- sleep 600

should either error (no name) or make up a name and set command to `["sleep", "600"]`.  It does neither, turning sleep into the pod name and making command 600.  `--` is supposed to define extra arguments.

### PR DESCRIPTION
**Pull request by feihujiang**
Fixes #13004

### CODE REVIEW

**Comment by feihujiang** [comment]
@smarterclayton @bgrant0607  Could you please review this PR?

File: pkg/kubectl/cmd/run.go [hunk 1]
```diff
@@ -106,6 +107,23 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
 		return cmdutil.UsageError(cmd, "NAME is required for run")
 	}
 
+	// Let kubectl run follow rules for `--`
+	var commandAndCustomArgs []string
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/run.go]: Hrm - this should be implemented by Cobra automatically.  Maybe I misunderstand what this is trying to fix?

Comment (smarterclayton) [inline pkg/kubectl/cmd/run.go]: Nm, I reminded myself.  There may be a more succint way to fix this

1. `kubectl run -p POD` should have zero args
2. `kubectl exec POD` should have zero args
3. `kubectl run POD COMMAND` should have 1 arg
4. `kubectl run -p POD COMMAND` should have 1 arg

If you use SetInterspersed on the cobra.Command object, does that give you the correct behavior in all of these cases?  (you might have to grab arg[0] for case 3 and shorten args to args[1:])

**Comment by feihujiang** [comment]
@smarterclayton  This commit fixes #13004 you posted, that is  "kubectl run --restart=Never --image=gcr.io/google_containers/busybox --command -- sleep 600
should either error (no name) or make up a name and set command to ["sleep", "600"]. It does neither,  turning sleep into the pod name and making command 600. "  I understood the bug is that  -- isn't effective, -- couldn't seperate between the name and <cmd> <arg1> ...  Did I get your point in #13004?

File: pkg/kubectl/cmd/run.go (hunk 1 above)
Comment (feihujiang) [inline pkg/kubectl/cmd/run.go]: This commit doesn't fix these cases, it fixes #13004 . And run command hasn't -p operation.

**Comment by feihujiang** [comment]
@k8s-bot  test this please

**Comment by eparis** [comment]
I'm trying to understand what we believe the 'syntax' is. These things work today:
1. `kubectl run --image=busybox --command mycontainer echo "hello there"`
2. `kubectl run --image=busybox --command mycontainer -- echo "hello there"`
3. `kubectl run --image=busybox --command -- mycontainer echo "hello there"`

But this patch will break 1 and 3...

**Comment by bgrant0607** [comment]
--command was added by #12571, which was a 1.1-alpha release, so backward compatibility is not a concern. We should fix the command to be what we want it to be before we cut 1.1 -- so ASAP.

**Comment by bgrant0607** [comment]
Could we just eliminate `--command` and just rely on `--`?

**Comment by bgrant0607** [comment]
cc @kubernetes/kubectl

**Comment by bgrant0607** [comment]
```
kubectl run mycontainer --image=busybox -- echo -e "\ahello there"
```

**Comment by bgrant0607** [comment]
Nevermind about --command. I see that's a bool. That's fine.
```
kubectl run mycontainer --image=busybox --command -- echo -e "\ahello there"
```

**Comment by feihujiang** [comment]
@bgrant0607 --command isn't problem, the problem is that -- is useless now, we can't separate NAME and (<cmd> <arg1> ... <argN>), for example, If miss mycontainer, kubectl run --image=busybox --command -- echo -e "\ahello there", echo would become the name.

**Comment by bgrant0607** [comment]
@eparis #13004 proposed that your (3) should not work. It's also fine by me if (1) does not work -- that we require `--`. But we should be consistent with exec:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/exec.go#L56

How does exec not suffer the same problem?

**Comment by feihujiang** [comment]
@bgrant0607 The exec has the same problem, I point it out in #13004

**Comment by feihujiang** [comment]
If we have the conclusion, I'd like to fix the same problem in exec.

**Comment by feihujiang** [comment]
@k8s-bot test this please

**Comment by feihujiang** [comment]
@k8s-bot test this again please

**Comment by eparis** [comment]
So `git log` (and other git like commands) don't allow 3) to work either. I don't mind if we 'break' that, the way that @feihujiang did.

I personally don't love breaking 1) and think it should continue to work. If the consensus is to force `--` all the time I won't jump up and down and scream and scream and scream.

@feihujiang seems like it might be useful if cobra stored len(args) when `--` was located. Let me write that quickly...

**Comment by eparis** [comment]
something like: https://github.com/eparis/pflag/commit/b25fea9a8363e6c84673209fa30d9e0023546599

**Comment by eparis** [comment]
along with https://github.com/eparis/cobra/commit/edde52e321465d7e0fbb1e1a32a8a8225ddd1f6a

**Comment by smarterclayton** [comment]
I don't like forcing -- either

On Sep 16, 2015, at 11:02 AM, Eric Paris <notifications@github.com> wrote:

So git log (and other git like commands) don't allow 3) to work either. I
don't mind if we 'break' that, the way that @feihujiang
<https://github.com/feihujiang> did.

I personally don't love breaking 1) and think it should continue to work.
If the consensus is to force -- all the time I won't jump up and down and
scream and scream and scream.

@feihujiang <https://github.com/feihujiang> seems like it might be useful
if cobra stored len(args) when -- was located. Let me write that quickly...

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13917#issuecomment-140768429>
.

**Comment by eparis** [comment]
My proposal:
```
- kubectl run myname echo "hello there"
Should be parsed NAME=myname; COMMAND='echo "hello there"'
- kubectl run myname -- echo "hello there"
Should be parsed NAME=myname; COMMAND='echo "hello there"'

- kubectl run -- myname echo "hello there"
Should be parsed NAME=''; COMMAND='myname echo "hello there"' (thus fail because no name)
- kubectl run echo "hello there"
Should be parsed NAME=echo; COMMAND='"hello there"' (thus fail because echo not valid)
```

Does that sound like good parsing and command syntax? @smarterclayton complain in #13004 is equivalent to my first "failed" case.

If we all agree on that syntax and behaviour, the code should be pretty easy.

**Comment by eparis** [comment]
I'm not sure how we should handle

```
kubectl run myname echo -- "hello there"
```

I know python has the concept of defined but unnamed args, i'll try to thing about how to do that in pflag/cobra, in which case "NAME" could make use of such a thing...

**Comment by smarterclayton** [comment]
> `kubectl run myname echo -- "hello there"`

Feels like `run POD=myname COMMAND=[echo, "--", "hello there"]` - after all, what if i wanted to do `kubectl run mypod git log -- somefile`

**Comment by smarterclayton** [comment]
Your proposal all looks like what I would expect as a regular human trying to make sense of this crazy world.

**Comment by eparis** [comment]
I'd contend:
`kubectl run myname -- echo -- "hello there"`
should pase as `COMMAND=["echo", "--", "hello there"]`

But `kubectl run myname echo -- "hello there"`
should parse as `COMMAND=["echo", "hello there"]`

**Comment by smarterclayton** [comment]
Yeah, makes sense.  Slightly surprising, but the first is what i would fall
back to if i got confused.

On Wed, Sep 16, 2015 at 1:05 PM, Eric Paris <notifications@github.com>
wrote:

> I'd contend:
> kubectl run myname -- echo -- "hello there"
> should pase as COMMAND=["echo", "--", "hello there"]
>
> But kubectl run myname echo -- "hello there"
> should parse as COMMAND=["echo", "hello there"]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13917#issuecomment-140805702>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by bgrant0607** [comment]
So is there a bug here that needs to be fixed, then?

**Comment by eparis** [comment]
I would say yes. This case: `kubectl run -- myname echo "hello there"` is not handled today how it should be. If @feihujiang agrees and says that my cobra/pflag additions would be helpful to a solution, I'd be happy to merge those tomorrow.

**Comment by feihujiang** [comment]
@eparis    The only question in your propoal: 
    - kubectl run echo "hello there"
    Should be parsed NAME=echo; COMMAND='"hello there"' (thus fail because echo not valid)
Why echo not valid? Does it fail because COMMAND=""hello there"" invalid?

I got others. And yes, your cobra/pflag additions is useful. So that we can use the function    ArgsLenAtDash() you added to find whether lack NAME.

**Comment by feihujiang** [comment]
@eparis Did you merge cobra/pflag additions and update that in kubernetes?

**Comment by feihujiang** [comment]
@smarterclayton @eparis @bgrant0607 I updated this commit based on your feedback, please review. kubectl run follow dash rules in this PR, and kubectl exec follow dash rules in #14161

**Comment by smarterclayton** [comment]
What is the --command in the usage statement?

**Comment by feihujiang** [comment]
--command default is false, If true and extra arguments are present, use them as the 'command' field in the container, rather than the 'args' field which is the default."

**Comment by feihujiang** [comment]
Kubernetes runs the image's default command. If we want to run a particular command or override the image's defaults, there are two additional fields that we can use:

Command: Controls the actual command run by the image
Args: Controls the arguments passed to the command

see https://github.com/kubernetes/kubernetes/blob/master/docs/user-guide/containers.md for more detail

**Comment by feihujiang** [comment]
@eparis Could you please review this PR?

**Comment by feihujiang** [comment]
@k8s-bot unit test this please

**Comment by feihujiang** [comment]
@eparis Could you please review this PR?  Thanks very much. 

I don't know why this builds timed-out on Shippable each time. Jenkins GCE e2e and Jenkins unit/integration all passed.

**Comment by janetkuo** [comment]
Restarting Shippable

**Comment by feihujiang** [comment]
I rebased, please review.

**Comment by feihujiang** [comment]
@k8s-teamcity-mesosphere test this please

**Comment by feihujiang** [comment]
@janetkuo Could you please review this PR?

**Comment by smarterclayton** [comment]
Sorry for the assign spam, bad github UI on mobile.

LGTM

**Comment by feihujiang** [comment]
@k8s-bot test this please

**Comment by feihujiang** [comment]
@k8s-bot test this please

**Comment by feihujiang** [comment]
I tested this PR locally, and all tests of unit/integration passed. I don't know why Jenkins unit/integration failing. Who could post the result of Jenkins unit/integration for me? I couldn't get it. Thanks very much.

**Comment by feihujiang** [comment]
I rebased, and found the error.

**Comment by janetkuo** [comment]
e2e failure:
```
22:44:38 Summarizing 1 Failure:
22:44:38 
22:44:38 [Fail] Probing container [It] with readiness probe should not be ready before initial delay and never restart [Conformance] 
22:44:38 /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/container_probe.go:57
```

unit/integration failure:
```
22:27:30 FAIL	k8s.io/kubernetes/pkg/kubectl/cmd [build failed]
...
22:30:48 !!! Error in ./hack/test-go.sh:239
22:30:48   'return ${rc}' exited with status 2
22:30:48 Call stack:
22:30:48   1: ./hack/test-go.sh:239 main(...)
22:30:48 Exiting with status 1
```

**Comment by feihujiang** [comment]
@smarterclayton I rebased this PR (use fake.CreateHTTPClient),  please review again. @janetkuo thanks.

**Comment by feihujiang** [comment]
@smarterclayton Could you please review this PR, this PR is delayed a lot.  Thanks!

**Pull request merged**